### ISSUE TITLE
BEEFY: add BEEFY-specific warp proofs so we can warp sync when using BEEFY

### ISSUE THREAD

**Issue by acatangiu** [paritytech/polkadot-sdk#1118]
Right now BEEFY doesn't work with warp sync.

We should add BEEFY warp proofs so that BEEFY can properly initialize when doing a warp sync.

We could then further enhance warp syncing to be able to sync exclusively using BEEFY warp-proofs. Light clients could warp sync without any GRANDPA proofs.

**Comment by Polkadot-Forum** [comment]
This issue has been mentioned on **Polkadot Forum**. There might be relevant details there:

https://forum.polkadot.network/t/warp-sync-and-beefy/4616/1

**Comment by Polkadot-Forum** [comment]
This issue has been mentioned on **Polkadot Forum**. There might be relevant details there:

https://forum.polkadot.network/t/polkadot-digest-30-nov-2023/5010/1

**Comment by acatangiu** [comment]
Discussed this with @andresilva and @serban300 and we came up with following plan:
- change GRANDPA warp sync protocol name from `/<genesis_hash>/sync/warp` to `/<genesis_hash>/sync/warp/grandpa` (while still supporting old name too),
- create new sync protocol `/<genesis_hash>/sync/warp/beefy/ecdsa` which will be a separate instance (aka both can run in parallel)
- switch substrate nodes to using `sync/warp/beefy/ecdsa` when doing `--sync warp` - nodes will warp sync using BEEFY
- deprecate but still support _providing_ GRANDPA WarpSyncProofs on `sync/warp/grandpa` protocol for light-clients (e.g. smoldot) while they (potentially) switch to BEEFY warp proofs

Later this year, when we get BLS support in BEEFY, create new protocol that will replace others (again with deprecation period): `/<genesis_hash>/sync/warp/beefy/bls`. Once we have BEEFY-BLS, warp proofs will be orders of magnitude smaller than current GRANDPA ones and everyone _will want_ to move to that.

cc @tomaka

**Comment by tomaka** [comment]
Sounds overall good to me, but I'm confused by:

> deprecate but still support providing GRANDPA WarpSyncProofs on sync/warp/grandpa protocol for light-clients (e.g. smoldot) while they (potentially) switch to BEEFY warp proofs

Is the idea that in the long term, you would do only a BEEFY warp sync, then download or get gossiped the Grandpa commits/justifications for the blocks between the BEEFY head and the GRANDPA head?

With the removal of the GRANDPA warp sync, the light client couldn't know anymore which block has a justification and which hasn't, which means that it would have to download every block between the two heads.
This isn't a problem by itself, but it might mean a longer initialization time, as the warp sync no longer propulsates you directly to where you want but a gap of unknown size behind.

Overall this increase in loading time would probably be compensated by the BLS signatures, but it's not totally clear to me that it's a good idea to get rid of GRANDPA warp syncing altogether.

> Later this year,

Do you mean 2023? If it's soon then I don't see the point in having EDCSA first? But I guess you mean 2024?

**Comment by nicolasochem** [comment]
Was this fixed? Should this be closed?

https://github.com/paritytech/polkadot-sdk/blob/master/prdoc/1.6.0/pr_2689.prdoc

**Comment by serban300** [comment]
No, we should keep this one opened. The issues are related, but they differ. The PR that you linked removed the limitation of not being able to run BEEFY when GRANDPA warp sync is enabled. This issue is about being able to use BEEFY warp sync proof instead of GRANDPA warp sync proofs. It would be an optimization.

**Comment by nicolasochem** [comment]
> No, we should keep this one opened. The issues are related, but they differ. The PR that you linked removed the limitation of not being able to run BEEFY when GRANDPA warp sync is enabled. This issue is about being able to use BEEFY warp sync proof instead of GRANDPA warp sync proofs. It would be an optimization.

Thanks, can you clarify though, whether warp-syncing a Kusama validator is ok on 1.0.7?

If I warp-sync a fresh kusama validator now, and rotate all keys (including beefy keys), will all validator duties be performed correctly?

**Comment by acatangiu** [comment]
> Thanks, can you clarify though, whether warp-syncing a Kusama validator is ok on 1.0.7?

Warp syncing a Kusama validator is ok on **1.7.0**.

> If I warp-sync a fresh kusama validator now, and rotate all keys (including beefy keys), will all validator duties be performed correctly?

Yes (as long as you use 1.6.x or 1.7.x node version to do the warp-sync - ideally clean DB too before).

**Comment by Polkadot-Forum** [comment]
This issue has been mentioned on **Polkadot Forum**. There might be relevant details there:

https://forum.polkadot.network/t/polkadot-digest-21-feb-2024/6358/1
